### PR TITLE
Enable detail view for any row

### PR DESCRIPTION
**Pull request by iethree**
## Background

Previously, object detail view was only available if
1. the table row had a visible primary key, OR
2. the table row was the only visible row due to filtering

This made object detail view inaccessible for many tables: e.g.
1. Tables with aggregations (e.g. 10 most popular products in the past 90 days)
2. Tables with hidden primary keys
3. Tables with composite primary keys (until you filtered down to a single record)
4. Tables based on SQL queries (where we can't detect primary keys)
5. Models without primary keys designated in metadata

### Redux 

The must-have-a-primary-key limitation is a product of the use of `zoomedRowId` in redux only reading from a map of primary keys. I have changed `zoomedRowId` from always being a primary key, to being a primary key if only one exists, but falling back to a row index if the dataset does not have exactly one primary key column.

Now, the `getPKRowIndexMap` selector will fall back to creating a simple map of row indexes instead of primary keys if there are no primary keys.

Because we can always fall back to row indices, now all tables (except pivot tables) are eligible to use object detail view.  (Except that we check if a table is an aggregation and hide object detail shortcuts for those. We can't do this for SQL queries).

### URLs

One potential pitfall of this behavior is that links for tables that do not have primary keys will be potentially unstable (e.g the 4th most popular product over the past 90 days will change based on when you view the link) - therefore we don't insert the object detail id into the URL when we're viewing a row without a usable primary key.

### Modal Titles

Part of the consequence of showing anything in a modal is we had to make our modal title logic a bit more complex.

Entity Name | Table Name | Question Name | Primary Key | First Row Data | Modal Title
-- | -- | -- | -- | -- | --
Sprocket | Widgets | Popular Widgets | 123 | 456 | Sprocket 123
  | Widgets | Popular Widgets | 123 | 456 | Widget 123
  |   | Popular Widgets | 123 | 456 | Popular Widget 123
  |   |   | 123 | 456 | Item Detail 123
  |   |   |   | 456 | Item Detail 456
Sprocket | Widgets | Popular Widgets |   | 456 | Sprocket
  | Widgets | Popular Widgets |   | 456 | Widget 456
  |   | Popular Widgets |   | 456 | Popular Widget 456

Screenshot from 2022-05-03 14-11-03

### CODE REVIEW

File: frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx [hunk 1]
```diff
@@ -3,8 +3,10 @@ import { isFK, isPK } from "metabase/lib/schema_metadata";
 import { zoomInRow } from "metabase/query_builder/actions";
 
 function hasManyPKColumns(question) {
-  const table = question.query().table();
-  const fields = table?.fields ?? question.getResultMetadata();
+  const fields = question.isDataset()
+    ? question.getResultMetadata() ?? question.query().table?.()?.fields
+    : question.query().table?.()?.fields ?? question.getResultMetadata();
```
Comment (iethree) [inline frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx:8]: possibly fixed by https://github.com/metabase/metabase/pull/22347 ?

Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/ObjectDetailDrill.jsx:8]: Unfortunately not, had to close the PR as the fix there was too fragile

**Review by kulyk — APPROVED** [review]
Nice work, pushed a few minor fixes

**Review by npfitz — APPROVED** [review]

**Review by mazameli — APPROVED** [review]
Nice!

**Pull request merged**